### PR TITLE
@alloy => [Sharing] Support sharing urls from inside martsy

### PR DESCRIPTION
**Pull request by orta**
Grabs article URLs before and offers the native share. 
Only supports Articles.

### CODE REVIEW

File: Artsy/Classes/Categories/Apple/NSString+StringBetweenStrings.h [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#import <Foundation/Foundation.h>
+
+@interface NSString (StringBetweenStrings)
```
Comment (orta) [inline Artsy/Classes/Categories/Apple/NSString+StringBetweenStrings.h]: This code is like nokogiri for me, it eventually ends up in every project.

https://github.com/orta/GIFs/blob/master/objc/GIFs/NSString+StringBetweenStrings.h

**Comment by orta** [comment]
Tests are in

File: Artsy/Classes/Categories/Apple/NSString+StringBetweenStrings.h (hunk 1 above)
Comment (alloy) [inline Artsy/Classes/Categories/Apple/NSString+StringBetweenStrings.h]: Might be implemented a bit simpler using `NSStringScanner` (another time)?

File: Artsy/Classes/Utils/ARSharingController.m [hunk 2]
```diff
@@ -75,9 +75,10 @@ - (void)presentActivityViewControllerFromButton:(UIButton *)button;
     // tell the caller we're done (from e.g. a delegate).
     //
     // TODO It appears that on iOS 8 the popver is retained by the system? In which case this can be removed.
-    if (popover) {
+    if (popover && [view isKindOfClass:UIButton.class]) {
```
Comment (alloy) [inline Artsy/Classes/Utils/ARSharingController.m]: Why are you only running this code when the view is a button? Not running this code (the runloop code) will prematurely deallocate the popover on iOS 7.

File: Artsy/Classes/View Controllers/ARInternalShareValidator.m [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+- (NSString *)addressBeingSharedFromShareURL:(NSURL *)url
+{
+    NSString *readableQuery = [url.query stringByReplacingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
+
+    if ([self isFacebookShareURL:url]) {
+        return [readableQuery componentsSeparatedByString:@"u="].lastObject;
+    } else if ([self isTwitterShareURL:url]) {
+        return [readableQuery componentsSeparatedByString:@"url="].lastObject;
+    }
```
Comment (alloy) [inline Artsy/Classes/View Controllers/ARInternalShareValidator.m]: Do you feel safe enough about martsy not adding any other GET parameters, which could break this?

**Comment by alloy** [comment]
I guess this is just considered to be part of the ‘adding magazine/shows’ change and thus doesn’t need its own CHANGELOG entry?

**Comment by alloy** [comment]
Other than those comments this looks very groovy :+1:

**Comment by orta** [comment]
I think we should get it in the beta_changelog

File: Artsy/Classes/Utils/ARSharingController.m (hunk 2 above)
Comment (orta) [inline Artsy/Classes/Utils/ARSharingController.m]: good point

**Comment by alloy** [comment]
Ah yeah, good point.

File: Artsy/Classes/View Controllers/ARInternalShareValidator.m (hunk 3 above)
Comment (orta) [inline Artsy/Classes/View Controllers/ARInternalShareValidator.m]: Hrm, yeah, might have to go all out and do a proper params dict + values.

**Comment by orta** [comment]
pleasantly found that NSURLComponents has that covered, https://gist.github.com/joemasilotti/09fe1f247a3da1c782dd

Code got a rewrite, passes locally.

**Comment by orta** [comment]
hrm, passes test,but doesn't work on device, testing

**Comment by orta** [comment]
bah, iOS8 only :/

@property (copy) NSArray *queryItems NS_AVAILABLE(10_10, 8_0);

**Comment by ashfurrow** [comment]
*cough cough* #285

**Comment by orta** [comment]
Yeah, so soon, https://github.com/artsy/eigen/issues/321#issuecomment-87352420

**Comment by orta** [comment]
For prosperity, incase someone improves this post iOS8
```
    NSURLComponents *components = [[NSURLComponents alloc] initWithURL:url resolvingAgainstBaseURL:YES];
    for (NSURLQueryItem *item in components.queryItems) {
        if ([self isFacebookShareURL:url] && [item.name isEqualToString:@"u"]) {
            return item.value;
        }
        if ([self isTwitterShareURL:url] && [item.name isEqualToString:@"url"]) {
            return item.value;
        }
    }
    return nil;
```

**Comment by orta** [comment]
OK, I'm fine with this for now 👍

**Pull request merged**